packages/api-client-core/src/GadgetConnection.ts

```diff
@@ -350,4 +321,51 @@ export class GadgetConnection {
 
     return headers;
   }
+
+  private async waitForOpenedConnection(
+    attempts: number,
+    timeoutMs: number,
+    overrides: GadgetSubscriptionClientOptions
+  ): Promise<SubscriptionClient> {
+    let subscriptionClient = this.newSubscriptionClient(overrides);
+    let unsubscribes: Function[] = []; // eslint-disable-line @typescript-eslint/ban-types
+
+    const clearListeners = () => {
+      unsubscribes.forEach((fn) => fn());
+      unsubscribes = [];
+    };
+
+    return await new Promise<SubscriptionClient>((resolve, reject) => {
+      const wrappedReject = (err: any) => {
+        if (err && err.code == CloseCode.ConnectionAcknowledgementTimeout && attempts > 0) {
+          attempts -= 1;
```

Comment:
Any reason you chose to do retries instead of repeating until the `setTimeout` fires? If the user is game waiting for N seconds overall, should we help them out by retrying until that timeout occurs? Or do you see it as "if we can't do it in N attempts, something bad is going on, just fail fast"?

Reply:
That's a good point. I added the `attempts` counter out of a general fear of any kind of infinite loop + hitting a service we control.

But the userland timeout should be good enough to stop this.